src/app/ngrx/reducers/episode-metrics.reducer.ts

```diff
@@ -1,32 +1,65 @@
-import { ActionTypes, CastleEpisodeMetricsAction } from '../actions';
-import { EpisodeMetricsModel, IntervalModel, MetricsType } from '../model';
-import { getMetricsProperty } from '../../shared/util/metrics.util';
+import { ActionTypes, AllActions, CastleEpisodeMetricsAction, CastleEpisodeChartToggleAction } from '../actions';
+import { MetricsType } from './metrics.type';
+import { IntervalModel } from './filter.reducer';
+import { getMetricsProperty } from './metrics.type';
+
+export interface EpisodeMetricsModel {
+  seriesId: number;
+  id: number;
+  guid?: string;
+  page?: number;
+  charted?: boolean;
+  monthlyDownloads?: any[][];
+  weeklyDownloads?: any[][];
+  dailyDownloads?: any[][];
+  hourlyDownloads?: any[][];
+}
 
 const initialState = [];
 
-export function EpisodeMetricsReducer(state: EpisodeMetricsModel[] = initialState, action: CastleEpisodeMetricsAction) {
-  let epIdx: number, episode: EpisodeMetricsModel, newState: EpisodeMetricsModel[];
+export function EpisodeMetricsReducer(state: EpisodeMetricsModel[] = initialState, action: AllActions) {
   switch (action.type) {
     case ActionTypes.CASTLE_EPISODE_METRICS:
-      const { id, seriesId, guid } = action.payload.episode;
-      epIdx = state.findIndex(e => e.seriesId === seriesId && e.id === id);
-      if (epIdx > -1) {
-        episode = {...state[epIdx], guid};
-        setEpisodeMetrics(action.payload.filter.interval, action.payload.metricsType, episode, action.payload.metrics);
-        newState = [...state.slice(0, epIdx), episode, ...state.slice(epIdx + 1)];
-      } else {
-        episode = { seriesId, id, guid };
-        setEpisodeMetrics(action.payload.filter.interval, action.payload.metricsType, episode, action.payload.metrics);
-        newState = [episode, ...state];
+      if (action instanceof CastleEpisodeMetricsAction) {
```

Comment:
I assume I'll learn from the Ultimate Angular videos why we're now checking instanceof? (Are we planning to use a single Action Type for multiple actions?) ... I should just watch the videos!

Reply:
I'm not sure you're gonna learn this from the videos. It's not initially straightforward or obvious to me at least, but the videos are setting up the action.payload to not have any additional levels of properties. If a new item is added to the state, say an episode, the episode's properties are on the action.payload rather than having an action.payload.episode at a child level. If it's an incoming response of a whole set of episodes, the payload is the array. They do this so that it can be assumed that payload is of type any. It seems like that eliminates the type checking benefits of having typed payloads. In order to follow the discriminating union pattern (our AllActions) we would need to be setting up our payloads similarly because of type safety. But if you want to have typed payloads, you can wrap these in instanceof checks to make TypeScript recognize the action and its payload type. I've had this open question about why the discriminating union pattern isn't working for a while; it's just that now I had to face it head on because I'm supporting more than one action type in some of the reducers. At least I had better examples to understand where I'd gotten off.